src/java/org/apache/cassandra/dht/RangeStreamer.java

```diff
@@ -353,9 +361,27 @@ public void addRanges(String keyspaceName, ReplicaCollection<?> replicas)
      */
     private boolean useStrictSourcesForRanges(AbstractReplicationStrategy strat)
     {
-        return useStrictConsistency
-                && tokens != null
-                && metadata.getSizeOfAllEndpoints() != strat.getReplicationFactor().allReplicas;
+        boolean res = useStrictConsistency && tokens != null;
+        
+        if (res)
+        {
+            int nodes = 0;
+
+            if (strat instanceof NetworkTopologyStrategy)
+            {
+                ImmutableMultimap<String, InetAddressAndPort> dc2Nodes = metadata.getDC2AllEndpoints(snitch);
+
+                NetworkTopologyStrategy ntps = (NetworkTopologyStrategy) strat;
+                for (String dc : dc2Nodes.keySet())
+                    nodes += ntps.getReplicationFactor(dc).allReplicas > 0 ? dc2Nodes.get(dc).size() : 0;
+            }
+            else
+                nodes = metadata.getSizeOfAllEndpoints();
+    
+            res = res && nodes > strat.getReplicationFactor().allReplicas;
```

Comment:
The `&&` part seems redundant here, since `res` is always true when evaluating this line, as it's shown by an IDE warning. We could just use:
```java
res = nodes > strat.getReplicationFactor().allReplicas;
```
Or maybe we could just eagerly return `false` after evaluating `useStrictConsistency` and `token` at the beginning of the method, and so also reduce nesting?
```java
if (!useStrictConsistency || tokens == null)
    return false;

int nodes = 0;

if (strat instanceof NetworkTopologyStrategy)
{
    ImmutableMultimap<String, InetAddressAndPort> dc2Nodes = metadata.getDC2AllEndpoints(snitch);
    NetworkTopologyStrategy ntps = (NetworkTopologyStrategy) strat;
    for (String dc : dc2Nodes.keySet())
        nodes += ntps.getReplicationFactor(dc).allReplicas > 0 ? dc2Nodes.get(dc).size() : 0;
}
else
{
    nodes = metadata.getSizeOfAllEndpoints();
}

return nodes > strat.getReplicationFactor().allReplicas;
```

Reply:
I was thinking about branch prediction when writing that. Iirc from some training an early return disables it so accumulating the result to return at the end is preferable. You'd then want to keep `nodes` declaration inside the branch to spare that cost when possible as well.

So I am removing the `res &&` bit indeed but the rest sgtm. Does it make sense or am I being too far reached? This is not in the hot path iiuc so we may want to favor readability :thinking:

Reply:
Since the function is not in the hot path I'd prefer readability (if the the early return is more readable, which is debatable), but I'm not against the current approach if you prefer it.